### ISSUE TITLE
neo4j, Wildfly, Transactions

### ISSUE THREAD

**Issue by spiewok** [neo4j/neo4j-jdbc#93]
Hi, 
I have a hard time to get neo4j working with transactions. I use BMT together with wildfly (also tried CMT). I call the business method in a stateless bean and have a DB facade injected via CDI. So the business method actually calls two save-methods in the facade. The code is as follows:

@StatelessBean

try {

                tx.begin();

                //saveSomething using pure JDBC

               //again save Something using pure JDBC

                tx.commit();

        } catch (Exception ex) {
            try {
                tx.rollback();
            } catch (SystemException e) {
                LOGGER.error(e.getMessage(),e);
            }
        }

Now, if an error happens, the next time a call the business method, the Prepared Statement fails with the following error message:
Caused by: org.neo4j.driver.v1.exceptions.ClientException: Cannot run more statements in this transaction, because previous statements in the transaction has failed and the transaction has been rolled back. Please start a new transaction to run another statement.

Actually, as you can see in the code, I'm starting a new transaction at the beginning of the method.

Why is that? Am I missing something? Is the driver not supposed to work in a JavaEE environment?

Could you please shed some light on this? Maybe it's a bug? Since I'm struggling with it for two days now, I could use any help that's available. Thanks!

Christian

**Comment by inserpio** [comment]
Hi @spiewok, which protocol are you using? HTTP? BOLT? Or have you tried with both?
And is AutoCommit On or Off?

**Comment by spiewok** [comment]
Hi @inserpio,
I'm using BOLT. The datasource is declared in the Wildfly config. I didn't try HTTP. AutoCommit is Off (my understanding is, that the container is doing this automatically as I use BMT). It is really strange. And it's just in case an error happens. So the sequence is:
1. Trigger the method. Another exception is raised, the transaction is rolled back correctly.
2. Trigger the method again. Now, the first call on a connection is creating the Transaction-Exception.
3. Trigger the method. Again, the other exception is raised, the transaction is rolled back.
2. Trigger the method again. Now, the first call on a connection is creating the Transaction-Exception.
....and so on....

I don't see that I'm doing something wrong....

Here is some more of the stacktrace. Maybe this helps:
Caused by: org.neo4j.driver.v1.exceptions.ClientException: Cannot run more statements in this transaction, because previous statements in the transaction has failed and the transaction has been rolled back. Please start a new transaction to run another statement.
	at org.neo4j.driver.internal.InternalTransaction.ensureNotFailed(InternalTransaction.java:189)
	at org.neo4j.driver.internal.InternalTransaction.run(InternalTransaction.java:158)
	at org.neo4j.driver.internal.InternalTransaction.run(InternalTransaction.java:132)
	at org.neo4j.driver.internal.InternalTransaction.run(InternalTransaction.java:145)
	at org.neo4j.jdbc.bolt.BoltPreparedStatement.executeInternal(BoltPreparedStatement.java:94)
	at org.neo4j.jdbc.bolt.BoltPreparedStatement.executeQuery(BoltPreparedStatement.java:54)
	at org.jboss.jca.adapters.jdbc.WrappedPreparedStatement.executeQuery(WrappedPreparedStatement.java:504)
	... 111 more

**Comment by spiewok** [comment]
Oh and I just tried (again) CMT. Same effect. If an transaction is set to rollback, the next call to the DB fails with the same exception.

So my solution for the moment is: no rollback.... :-(

**Comment by inserpio** [comment]
The transaction is (correctly) marked as rollback, but it looks like a new one is not created from the second call and so on... How do you get the `tx` reference? Can you post the complete code fragment? Let me investigate a bit the code and yes it could be a bug...

**Comment by inserpio** [comment]
```
	@Override public void commit() throws SQLException {
		this.checkClosed();
		this.checkAutoCommit();
		if (this.transaction == null) {
			throw new SQLException("The transaction is null");
		}
		this.transaction.success();
		this.transaction.close();
		this.transaction = this.session.beginTransaction();
	}
	@Override public void rollback() throws SQLException {
		this.checkClosed();
		this.checkAutoCommit();
		if (this.transaction == null) {
			throw new SQLException("The transaction is null");
		}
		this.transaction.failure();
	}
```
We don't renew the transaction in the `rollback` method as we do for `commit`.
I wanna try to see if this could the cause.

**Comment by spiewok** [comment]
Yes, that was my impression too. Did just change everything back to CMT and I have the same behaviour.

First the code for CMT:

@Stateless
public class SignUpPersistorBean {
    private static final Logger LOGGER = LoggerFactory.getLogger(SignUpPersistorBean.class);

    @Inject
    PasswordProvider passProvider;

    @EJB
    PracticeBeanLocal practiceBeanLocal;

    @Resource
    private EJBContext context;


    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
    public void saveAllData(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        try {

			//Here are the calls to other business methods from a second EJB.
			//They are not annotated which basically means TransactionAttributType.REQUIRED
            VPractice practice = createPractice(request);
            VUserProvider userProvider = createUserProvider(request);

            response.sendRedirect("/sign_up_finish.jsp");
        } catch (Exception ex) {
            context.setRollbackOnly();
            request.getSession().setAttribute("error", ex);
            response.sendRedirect("/sign-up-error.jsp");
        }

    }
}




and here is the code for BMT:




@Stateless
@TransactionManagement(TransactionManagementType.BEAN)
public class SignUpPersistorBean {
    private static final Logger LOGGER = LoggerFactory.getLogger(SignUpPersistorBean.class);

    @Inject
    PasswordProvider passProvider;

    @Inject
    MailSender mailSender;

    @EJB
    PracticeBeanLocal practiceBeanLocal;

    @Resource
    private UserTransaction tx;

    public void saveAllData(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        try {

            tx.begin();

            VPractice practice = createPractice(request);
            VUserProvider userProvider = createUserProvider(request);

            tx.commit();

            response.sendRedirect("/sign_up_finish.jsp");
        } catch (Exception ex) {
            try {
                tx.rollback();
            } catch (SystemException e) {
                LOGGER.error(e.getMessage(), e);
            }
            request.getSession().setAttribute("error", ex);
            response.sendRedirect("/sign-up-error.jsp");
        }

    }


the business methods are really just plain JDBC calls without any transaction annotation or call.

**Comment by inserpio** [comment]
@spiewok could you try with this jar?
[neo4j-jdbc-bolt-3.0.2.jar.zip](https://github.com/neo4j-contrib/neo4j-jdbc/files/797650/neo4j-jdbc-bolt-3.0.2.jar.zip)

**Comment by spiewok** [comment]
Hi @inserpio,
thank you very, very much for answering so fast!

I did actually download the sources yesterday already and did make the proposed changes by myself and testet it and so far I tested everything and it's working. It was indeed missing the creation of the new transaction. So everything is working perfectly now! (I didn't try the jar you sent me, since everything is working as excepted now).

Thanks again, problem solved!
Christian
